File: problems/0701-0800/0727_longest-shared-leading-digits/solution_sorted_bracketing.ts
function longestSharedPrefix(arr1: number[], arr2: number[]): number {
    // The deepest cross-array agreement is realized by two lexicographically
    // adjacent entries, so merge both arrays as source-tagged digit strings.
    const entries: [string, number][] = [];
    for (const x of arr1) entries.push([String(x), 0]);
    for (const y of arr2) entries.push([String(y), 1]);
    // Sort as digit strings, never numerically: only lexicographic order
    // keeps a prefix family in one contiguous block.
    entries.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
    let best = 0;
    for (let i = 1; i < entries.length; i++) {
        const [u, su] = entries[i - 1];
        const [v, sv] = entries[i];
        // Same-source neighbors cannot witness a cross pair.
        if (su === sv) continue;
        let shared = 0;
        for (let j = 0; j < u.length && j < v.length; j++) {
            if (u[j] !== v[j]) {
                // Digits diverge: the run cannot extend past here.
                break;
            }
            shared++;
        }
        if (shared > best) best = shared;
    }
    return best;
}

File: problems-adapt/3001-3100/3043_longest-shared-leading-digits/solution_sorted_bracketing.rs
impl Solution {
    pub fn longest_shared_prefix(arr1: Vec<i32>, arr2: Vec<i32>) -> i32 {
        // The deepest cross-array agreement is realized by two lexicographically
        // adjacent entries, so merge both arrays as source-tagged digit strings.
        let mut entries: Vec<(String, u8)> = Vec::with_capacity(arr1.len() + arr2.len());
        for &x in &arr1 {
            entries.push((x.to_string(), 0));
        }
        for &y in &arr2 {
            entries.push((y.to_string(), 1));
        }
        // Sort as digit strings, never numerically: only lexicographic order
        // keeps a prefix family in one contiguous block.
        entries.sort_by(|a, b| a.0.cmp(&b.0));
        let mut best: i32 = 0;
        for i in 1..entries.len() {
            // Same-source neighbors cannot witness a cross pair.
            if entries[i - 1].1 == entries[i].1 {
                continue;
            }
            let (u, v) = (&entries[i - 1].0, &entries[i].0);
            let mut shared: i32 = 0;
            for j in 0..u.len().min(v.len()) {
                if u.as_bytes()[j] != v.as_bytes()[j] {
                    // Digits diverge: the run cannot extend past here.
                    break;
                }
                shared += 1;
            }
            if shared > best {
                best = shared;
            }
        }
        best
    }
}
